File: cdb_subscription/iter_python/cdbl.py
# *********************************************************************
# ConfD CDB iteration example - python version
#
# (C) 2007-2017 Tail-f Systems
# Permission to use this code as a starting point hereby granted
# This is ConfD Sample Code.
#
# See the README file for more information
# ********************************************************************

import logging
import select
import socket
import sys

import _confd
import _confd.cdb as cdb
import _confd.maapi as maapi

import root_ns

confd_debug_level = _confd.TRACE
log_level = logging.INFO
CONFD_ADDR = '127.0.0.1'

logging.basicConfig(
    format="%(asctime)s:%(relativeCreated)s"
           "%(levelname)s:%(filename)s:%(lineno)s:%(funcName)s  %(message)s",
    level=log_level)
log = logging.getLogger("cdbl")


class Rfhead:
    sector_id = ''
    children = {}  # key is cdn, value is childattr


rfheads = {}  # key is dn, value is Rfhead


# helper function for logging (alternative to dump_db in C example)
def _str_rfh(rfheads):
    s = "{"
    for k1, v1 in rfheads.items():
        s += "%d:sector_id=%s," % (k1, v1.sector_id)
        s += " children: {"
        for k2, v2 in v1.children.items():
            s += " %d:childattr=%s, " % (k2, v2)
        s += "}, "
    s += "}"
    return s


def read_head(cdbsock, headkey):
    log.debug(
        "==> cdbsock.fileno()=%d headkey=%d" % (cdbsock.fileno(), headkey))
    hp = rfheads.get(headkey)

    if hp is None:
        hp = Rfhead()
        rfheads[headkey] = hp
        log.debug("Created new hp=%s" % hp)

    cdb.cd(cdbsock, "/root/NodeB/RFHead{%d}" % headkey)
    hp.sector_id = str(cdb.get(cdbsock, "SECTORID_ID"))
    hp.children.clear()
    log.debug("hp=%s", hp)
    n = cdb.num_instances(cdbsock, "Child")
    log.debug("number of child instances n=%d" % n)
    for i in range(n):
        dn = int(cdb.get(cdbsock, "Child[%d]/cdn" % i))
        childattr = str(cdb.get(cdbsock, "Child[%d]/childAttr" % i))
        hp.children[dn] = childattr

    log.debug("<== processed hp=%s" % hp)


def read_db(cdbsock):
    log.debug("==> cdbsock.fileno()=%d" % cdbsock.fileno())
    cdb.start_session(cdbsock, cdb.RUNNING)
    cdb.set_namespace(cdbsock, root_ns.ns.hash)
    n = cdb.num_instances(cdbsock, "/root/NodeB/RFHead")
    for k, v in rfheads.items():
        v.children.clear()
    rfheads.clear()
    for i in range(n):
        key = int(cdb.get(cdbsock, "/root/NodeB/RFHead[%d]/dn" % i))
        read_head(cdbsock, key)

    cdb.end_session(cdbsock)
    log.debug("<== rfheads=%s" % _str_rfh(rfheads))


def iter(kp, op, oldv, newv, cdbsock):
    log.info("=> kp=%s op=%i oldv=%s newv=%s cdbsock.fileno()=%d"
             % (kp, op, newv, oldv, cdbsock.fileno()))
    log.debug("rfheads=%s" % _str_rfh(rfheads))
    rv = _confd.ITER_CONTINUE
    if op == _confd.MOP_CREATED:
        log.debug("MOP_CREATED")
        log.info("Create: %s" % kp)
        ctag = kp[1]
        if ctag.tag == root_ns.ns.root_RFHead:
            log.debug("RFHead")
            # an rfhead was created
            # keypath is /root/NodeB/RFHead{$key}
            #              3     2      1    0
            read_head(cdbsock, int(kp[0][0]))  # kp[0] is a tuple of keys
        elif ctag.tag == root_ns.ns.root_Child:
            log.debug("Child")
            # a child to en existing rfhead was created.
            # keypath is /root/NodeB/RFHead{$key}/Child{$key2}
            #              5      4      3    2      1    0
            # we can here choose to read the  new child or reread
            # the entire head structure
            read_head(cdbsock, int(kp[2][0]))  # kp[2] is a tuple of keys

    elif op == _confd.MOP_DELETED:
        log.debug("MOP_DELETED")
        log.info("Delete: %s" % kp)
        dtag = kp[1]
        if dtag.tag == root_ns.ns.root_RFHead:
            log.debug("RFHead")
            # an rfhead was created
            # keypath is /root/NodeB/RFHead{$key}
            #              3     2      1    0
            headkey = int(kp[0][0])  # kp[0] is a tuple of keys
            log.debug("headkey=%d" % headkey)
            if rfheads.has_key(headkey):
                del rfheads[headkey]
            else:
                log.warn(
                    "Cannot find %d in rfheads map %s" % (
                        headkey, _str_rfh(rfheads)))
                rv = _confd.ITER_RECURSE
        elif dtag.tag == root_ns.ns.root_Child:
            log.debug("Child")
            # a child of an existing head was removed
            # keypath is /root/NodeB/RFHead{$key}/Child{$key2}
            #               5      4      3    2      1    0
            # we can here choose to read the  new child or reread
            # the entire head structure
            headkey = int(kp[2][0])
            childkey = int(kp[0][0])
            log.debug("headkey=%d childkey=%d" % (headkey, childkey))
            if rfheads.has_key(headkey):
                rh = rfheads[headkey]
                if rh.children.has_key(childkey):
                    del rfheads[headkey].children[childkey]
                else:
                    log.warn("Cannot find %d in children" % childkey)
            else:
                log.warn(
                    "Cannot find %d in rfheads map %s" % (headkey,
                                                          _str_rfh(rfheads)))
                rv = _confd.ITER_RECURSE

    elif op == _confd.MOP_MODIFIED:
        log.debug("MOP_MODIFIED")
        log.info("Modified %s" % kp)
        rv = _confd.ITER_RECURSE

    elif op == _confd.MOP_VALUE_SET:
        log.debug("MOP_VALUE_SET")
        log.info("Value Set: %s --> (%s)" % (kp, newv))
        leaf = kp[0]
        if leaf.tag == root_ns.ns.root_SECTORID_ID:
            log.debug("SECTOR_ID")
            # keypath is /root/NodeB/RFHead{$key}/SECTORID_ID
            #              4     3      2     1       0
            headkey = int(kp[1][0])  # kp[0] is a tuple of keys
            log.debug("headkey=%d" % headkey)
            if rfheads.has_key(headkey):
                rfheads[headkey].sector_id = str(newv)
                rv = _confd.ITER_RECURSE
            else:
                log.warn(
                    "Cannot find %d in rfheads map %s" % (
                        headkey, _str_rfh(rfheads)))
        elif leaf.tag == root_ns.ns.root_childAttr:
            log.debug("root_childAttr")
            # keypath is /root/NodeB/RFHead{$key}/Child{$key2}/childAttr
            #             6      5      4    3      2     1       0
            headkey = int(kp[3][0])
            childkey = int(kp[1][0])
            log.debug("headkey=%d childkey=%d" % (headkey, childkey))
            if rfheads.has_key(headkey):
                rh = rfheads[headkey]
                if rh.children.has_key(childkey):
                    rfheads[headkey].children[childkey] = str(newv)
                else:
                    log.warn("Cannot find %d in children rfheads map %s"
                             % (childkey, _str_rfh(rfheads)))
            else:
                log.warn("Cannot find %d in rfheads map %s"
                         % (headkey, _str_rfh(rfheads)))
    else:
        log.warn("Unexpected op %d for %s" % (op, kp))
        rv = _confd.ITER_RECURSE

    log.info("<== rv=%i rfheads=%s", rv, _str_rfh(rfheads))
    return rv


def run():
    log.info("==>")
    # In C we use confd_init() which sets the debug-level, but for Python the
    # call to confd_init() is done when we do 'import confd'.
    # Therefore we need to set the ConfD debug level here (if we want it to be
    # different from the default debug level - CONFD_SILENT):
    _confd.set_debug(confd_debug_level, sys.stderr)

    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
    subsock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
    # maapi socket for load schemas
    maapisock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, 0)
    maapi.connect(maapisock, CONFD_ADDR, _confd.CONFD_PORT)
    maapi.load_schemas(maapisock)

    cdb.connect(sock, cdb.DATA_SOCKET, CONFD_ADDR, _confd.CONFD_PORT)
    sub_path = "/root/NodeB/RFHead"
    cdb.connect(subsock, cdb.SUBSCRIPTION_SOCKET, CONFD_ADDR, _confd.CONFD_PORT)
    spoint = cdb.subscribe(subsock, 3, root_ns.ns.hash, sub_path)
    cdb.subscribe_done(subsock)
    log.debug("Subscribed to path %s spoint=%i" % (sub_path, spoint))
    read_db(sock);

    try:
        _r = [subsock]
        _w = []
        _e = []
        log.debug("subscok connected, starting ConfD loop")
        while (True):
            (r, w, e) = select.select(_r, _w, _e, 1)
            # log.debug("select triggered r=%r" % r)
            for rs in r:
                log.debug("rs.fileno=%i subscok.fileno=%i" % (
                    rs.fileno(), subsock.fileno()))
                if rs.fileno() == subsock.fileno():
                    log.debug("subsock triggered")
                    try:
                        sub_points = cdb.read_subscription_socket(subsock)
                        for s in sub_points:
                            if s == spoint:
                                log.debug("our spoint=%i triggered" % spoint)
                                cdb.start_session(sock, cdb.RUNNING)
                                cdb.set_namespace(sock, root_ns.ns.hash)
                                cdb.diff_iterate(subsock, spoint, iter,
                                                 _confd.ITER_WANT_PREV, sock)
                                cdb.end_session(sock)

                                # variant with diff_match not yet available
                                # in python

                                cdb.sync_subscription_socket(subsock,
                                                             cdb.DONE_PRIORITY)
                    except (_confd.error.Error) as e:
                        if e.confd_errno is not _confd.ERR_EXTERNAL:
                            raise e

    except KeyboardInterrupt:
        print("\nCtrl-C pressed\n")
    finally:
        sock.close()
        subsock.close()
    log.info("<==")


if __name__ == "__main__":
    log.info("==>")
    run()
    log.info("<==")

File: cdb_subscription/iter_java/Cdbl.java

import java.io.IOException;
import java.net.Socket;
import java.util.EnumSet;

import org.apache.log4j.Logger;

import com.tailf.cdb.Cdb;
import com.tailf.cdb.CdbDBType;
import com.tailf.cdb.CdbDiffIterate;
import com.tailf.cdb.CdbSession;
import com.tailf.cdb.CdbSubscription;
import com.tailf.cdb.CdbSubscriptionSyncType;
import com.tailf.cdb.CdbSubscriptionType;
import com.tailf.conf.Conf;
import com.tailf.conf.ConfBuf;
import com.tailf.conf.ConfInt64;
import com.tailf.conf.ConfKey;
import com.tailf.conf.ConfObject;
import com.tailf.conf.ConfPath;
import com.tailf.conf.ConfTag;
import com.tailf.conf.ConfValue;
import com.tailf.conf.DiffIterateFlags;
import com.tailf.conf.DiffIterateOperFlag;
import com.tailf.conf.DiffIterateResultFlag;

public class Cdbl {
    private static Logger log = Logger.getLogger(Cdbl.class);

    private static String confdAddress = "127.0.0.1";
    private static int confdPort = Conf.PORT;
    private static int headPoint;
    public final static int MAXH = 3;
    public final static int MAXC = 2;
    public static RFHead[] rfheads = new RFHead[MAXH];

    protected static class RFHead {
        protected ConfInt64 dn;
        protected ConfBuf sector_id;
        protected boolean inuse;
        protected Child[] children = new Child[Cdbl.MAXC];

        public RFHead() {
            for ( int i = 0; i < Cdbl.MAXC; i++ ) children[i] = new Child();
        }
    }

    protected static class Child {
        protected ConfInt64 dn;
        protected String childattr;
        protected boolean inuse;
    }


    public static void main(String arg[]) {
        if (arg.length > 0) {
            try {
                confdPort = Integer.parseInt(arg[1]);
            } catch (NumberFormatException e) {
                confdPort = Conf.PORT;
            }
        }

        Socket sock = null;
        try {
            sock = new Socket(confdAddress, confdPort);
        } catch ( IOException e ) {
            log.error("Could not open SocketChannel ");
            e.printStackTrace();
            System.exit(1);
        }


        Cdb cdb = null;
        Cdb cdb2 = null;
        CdbSubscription cdbSubscriber = null;
        try {
            cdb = new Cdb("Subscription-Sock",  sock);
            cdbSubscriber = cdb.newSubscription();
            headPoint =
                cdbSubscriber.subscribe(CdbSubscriptionType.SUB_RUNNING,
                                        3, root.hash, "/root/NodeB/RFHead");
            cdbSubscriber.subscribeDone();
            cdb2 = new Cdb(Cdbl.class.getName(), new Socket("127.0.0.1",
                                                            confdPort));
        } catch (Exception e) {
            log.error( "Could not connect to cdb", e);
            System.exit(1);
        }

        readDb(cdb2) ;
        dumpDb();

        try {
            while (true) {
                try {
                    int[] point = cdbSubscriber.read();
                    CdbSession cdbSession =
                        cdb2.startSession(CdbDBType.CDB_RUNNING);
                    EnumSet<DiffIterateFlags> diffFlags =
                            EnumSet.of(DiffIterateFlags.ITER_WANT_PREV);
                    cdbSubscriber.diffIterate(point[0], new DiffIterateImpl(),
                                              diffFlags, cdbSession);
                    cdbSession.endSession();
                } finally {
                    cdbSubscriber.sync(CdbSubscriptionSyncType.DONE_PRIORITY);
                    dumpDb ();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void dumpDb() {
        System.err.printf("\nDumping \n");
        for (int i = 0; i < MAXH; i++) {
            if (!Cdbl.rfheads[i].inuse) {
                continue;
            }
            System.err.printf("HEAD %d <%s>\n", rfheads[i].dn.longValue(),
                              rfheads[i].sector_id);
            for (int j = 0; j < Cdbl.MAXC; j++) {
                if (!Cdbl.rfheads[i].children[j].inuse) {
                    continue;
                }
                System.err.printf("    Child %d <<%s>> \n",
                                  Cdbl.rfheads[i].children[j].dn.longValue(),
                                  Cdbl.rfheads[i].children[j]
                                  .childattr.toString());
            }
        }

        System.err.printf(  "---------- \n");
    }

    static void readDb(Cdb cdb) {
        CdbSession cdbSession = null;
        try {
            cdbSession = cdb.startSession(CdbDBType.CDB_RUNNING);
            cdbSession.setNamespace(new root());
        } catch (Exception e) {
            log.error("Could not start new session ", e);
            return;
        }

        try {
            int n = cdbSession.getNumberOfInstances(
                                          new ConfPath("/root/NodeB/RFHead"));
            for (int i = 0; i < Cdbl.MAXH ; i++) {
                rfheads[i] = new RFHead();
            }

            for (int i = 0; i < n; i++) {
                ConfValue key =
                    cdbSession.getElem ( "/root/NodeB/RFHead[%d]/dn", i);
                readHead (cdbSession, new ConfKey(key));
            }
            cdbSession.endSession();
        } catch (Exception e) {
            e.printStackTrace();
            return;
        }
    }

    static void readHead(CdbSession cdbSession, ConfKey key) {
        int i = 0;
        int pos = -1;

        for (i = 0; i < Cdbl.MAXH; i++) {
            if (key.equals(new ConfKey(Cdbl.rfheads[i].dn))) {
                pos = i;
                break;
            }
        }

        if (pos == -1) {
            for (i = 0; i < MAXH; i++) {
                if (!rfheads[i].inuse) {
                    pos = i;
                    break;
                }
            }
        }
        System.err.printf("Picking %d \n" , pos) ;
        Cdbl.RFHead hp = Cdbl.rfheads[pos];

        try {
            cdbSession.cd("/root/NodeB/RFHead{%x}", key);
            hp.dn = (ConfInt64)cdbSession.getElem(new ConfPath("dn"));
            hp.sector_id =
                (ConfBuf) cdbSession.getElem( new ConfPath ("SECTORID_ID"));
            hp.inuse = true;
            int n = cdbSession.getNumberOfInstances("Child") ;

            for (i =0; i < n; i++) {
                hp.children[i].dn =
                    (ConfInt64) cdbSession.getElem("Child[%d]/cdn",i);
                hp.children[i].childattr =
                    ((ConfBuf) cdbSession.getElem("Child[%d]/childAttr",i)).
                                                                    toString();
                hp.children[i].inuse = true;
            }
        } catch (Exception e) {
            log.error(" Could not read values ", e);
        }
    }
}


class DiffIterateImpl implements CdbDiffIterate {
    private static Logger log = Logger.getLogger(DiffIterateImpl.class);
        public DiffIterateResultFlag iterate(ConfObject[] kp,
                                             DiffIterateOperFlag op,
                                             ConfObject oldValue,
                                             ConfObject newValue,
                                             Object initstate) {
            CdbSession cdbSession = (CdbSession) initstate;
            ConfPath path = new ConfPath(kp);
            switch (op) {
            case MOP_CREATED: {
                System.err.printf("Create: %s\n", path.toString());
                ConfTag tag = (ConfTag) kp[1];
                switch (tag.getTagHash()) {
                case root.root_RFHead: {
                    /* an rfhead was created */
                    /* keypath is /root/NodeB/RFHead{$key}  */
                    /*              3     2      1    0     */
                    /*       /ConfTag/ConfTag/ConfTag/ConfKey */
                    Cdbl.readHead(cdbSession, (ConfKey)kp[0]);
                    return DiffIterateResultFlag.ITER_CONTINUE;
                }
                case root.root_Child: {
                    /* a child to en existing rfhead was created */
                    /* keypath is /root/NodeB/RFHead{$key}/Child{$key2}  */
                    /*              5      4      3    2      1    0     */
                    /* we can here choose to read the  new child or reread */
                    /* the entire head structure       */
                    Cdbl.readHead(cdbSession, (ConfKey)kp[2]);
                    return DiffIterateResultFlag.ITER_CONTINUE;
                }
                }
            }
            case MOP_DELETED: {
                System.err.printf("Create: %s\n", path.toString());
                ConfTag tag = (ConfTag) kp[1];
                switch (tag.getTagHash()) {
                case root.root_RFHead: {/* an rfhead was deleted */
                    /* keypath is /root/NodeB/RFHead{$key}  */
                    /*              3     2      1    0     */
                    /*           /ConfTag/ConfTag/ConfTag/ConfKey */
                    ConfKey headkey = (ConfKey)kp[0];
                    int headpos = 0;
                    while (headpos < Cdbl.MAXH) {
                        Cdbl.RFHead hp = Cdbl.rfheads[headpos++];
                        if (hp.inuse && new ConfKey(hp.dn).equals(headkey)) {
                            for (int i = 0; i < Cdbl.MAXC; i++) {
                                hp.children[i].inuse = false;
                            }
                            hp.inuse = false;
                            return DiffIterateResultFlag.ITER_CONTINUE;
                        }
                    }
                    break;
                }
                case root.root_Child: {
                    /* a child of an existing head was removed */
                    /* keypath is /root/NodeB/RFHead{$key}/Child{$key2}  */
                    /*              5      4      3    2      1    0     */
                    /* we can here choose to read the  new child or reread */
                    /* the entire head structure                           */
                    ConfKey headkey = (ConfKey)kp[2];
                    ConfKey childkey = (ConfKey)kp[0];
                    /* Now given the key here, identifying an rfhead */
                    /* we find the rfhead which contains our child */
                    int headpos = 0;

                    while (headpos < Cdbl.MAXH) {
                        Cdbl.RFHead hp = Cdbl.rfheads[headpos++];
                        log.info("hp.dn:" + hp.dn);
                        log.info("headkey:" + headkey);
                        if (new ConfKey(hp.dn).equals(headkey)) {
                            int cpos = 0;
                            log.info(" inside hp.dn :" +hp.dn );
                            while (cpos < Cdbl.MAXC) {
                                log.info(" child cpos:" + cpos);
                                Cdbl.Child cp = hp.children[cpos++];
                                log.info(" child cp.dn:" + cp.dn);
                                log.info(" childkey:" + childkey);

                                if (childkey.equals(new ConfKey(cp.dn))) {
                                    log.info(" equals!");
                                    cp.inuse = false;
                                    return DiffIterateResultFlag.ITER_CONTINUE;
                                }
                            }
                        }
                    }
                }
                }
            }
            case MOP_MODIFIED: {
                System.err.printf( "Modified %s\n", path.toString());
                break;
            }
            case MOP_VALUE_SET: {
                System.err.printf("Value Set %s --> %s \n",
                                  path.toString(), newValue);
                ConfTag tag = (ConfTag) kp[0];
                switch (tag.getTagHash()) {
                case root.root_SECTORID_ID: {
                    /* keypath is /root/NodeB/RFHead{$key}/SECTORID_ID  */
                    /*              4     3      2     1       0        */
                        ConfKey headkey = (ConfKey) kp[1];
                        int headpos = 0;

                        while (headpos < Cdbl.MAXH) {
                            Cdbl.RFHead hp = Cdbl.rfheads[headpos++];
                            if (headkey.equals(new ConfKey(hp.dn))) {
                                hp.sector_id  = (ConfBuf) newValue;
                                return DiffIterateResultFlag.ITER_RECURSE;
                            }
                        }
                    }
                    break;
                    case root.root_childAttr: {
          /* keypath is /root/NodeB/RFHead{$key}/Child{$key2}/childAttr  */
          /*              6      5      4    3      2     1       0      */
                        ConfKey headkey = (ConfKey) kp[3];
                        ConfKey childkey = (ConfKey) kp[1];
                        int headpos = 0;
                        while (headpos < Cdbl.MAXH) {
                            Cdbl.RFHead hp = Cdbl.rfheads[headpos++];
                            if (new ConfKey(hp.dn).equals(headkey)) {
                                int cpos = 0;
                                while (cpos < Cdbl.MAXC) {
                                    Cdbl.Child cp = hp.children[cpos++];
                                    if (new ConfKey(cp.dn).equals(childkey)) {
                                        cp.childattr = newValue.toString();
                                        return
                                            DiffIterateResultFlag.ITER_CONTINUE;
                                    }
                                }
                            }
                        }
                    }
                    }
                break;
            }
            default:
                System.err.printf("Unexpected op %d for %s\n", op,
                                  path.toString());
            }
            return DiffIterateResultFlag.ITER_RECURSE;
        }
}


